### DISCUSSION TITLE
Why the tensor has been changed when I change the array? I'm confused!

### DISCUSSION THREAD

**Discussion by Javidhassani** [mrdbourke/pytorch-deep-learning#124] [Q&A] [answered]
# Converting an array to a tensor

### Input:
```
green_array = np.arange(1, 8, dtype="float32")
numpy2tensor = torch.from_numpy(green_array)

green_array[2] = 20
green_array, numpy2tensor
```
### Output:
```
(array([ 1.,  2., 20.,  4.,  5.,  6.,  7.], dtype=float32),
 tensor([ 1.,  2., 20.,  4.,  5.,  6.,  7.]))

```

### But in this case, everything is okay:

**Input:**
```
green_array = np.arange(1, 8, dtype="float32")
numpy2tensor = torch.from_numpy(green_array)

green_array = green_array + 4
green_array, numpy2tensor
```

**Output:**
```
(array([ 5.,  6.,  7.,  8.,  9., 10., 11.], dtype=float32),
 tensor([1., 2., 3., 4., 5., 6., 7.]))
```

### And again, the tensor has been changed when I change the array :
**Input:**

```
green_array = np.arange(1, 8, dtype="float32")
numpy2tensor = torch.from_numpy(green_array)

green_array += 4
green_array, numpy2tensor
```
**Output:**
```
(array([ 5.,  6.,  7.,  8.,  9., 10., 11.], dtype=float32),
 tensor([ 5.,  6.,  7.,  8.,  9., 10., 11.]))
```

**Comment by AlessandroMiola** [comment, marked as answer]
I had the same doubt when going through the video; the point should be the following:

- On the one hand, as the [documentation](https://pytorch.org/docs/stable/generated/torch.from_numpy.html) states 
> The returned tensor and ndarray share the same memory. Modifications to the tensor will be reflected in the ndarray and vice versa.

i.e. modifications to the numpy array from which the torch tensor is created get reflected into the tensor itself (as in your first and third example).

- On the other hand, as explained in https://stackoverflow.com/questions/48370286/pytorch-updating-numpy-array-not-updating-the-corresponding-tensor, in your second example you're creating a brand new array, which is why the update doesn't get reflected in the torch tensor. For further understanding, please have a look at this snippet, too:

        arr = np.arange(0., 10.)
        print(f'Id of arr is {id(arr)}')
        arr += 1
        print(f'Id of arr is {id(arr)}')
        arr = arr + 1
        print(f'Id of arr is {id(arr)}')
